File: src/tax/us/schedule_d.py
"""
IRS Schedule D summary generator - standalone functions for tax module system.

Adapted from src/reports/schedule_d.py to work with Disposal objects directly.

Schedule D aggregates the totals from Form 8949:
  - Part I:  Short-term capital gains/losses
  - Part II: Long-term capital gains/losses
  - Part III: Net capital gain/loss + carryover calculation

Capital loss carryover rules:
  - Net capital loss limited to $3,000/year deduction against ordinary income
  - Excess carries over indefinitely to future years
"""
from __future__ import annotations

from decimal import Decimal

from src.tax.models import HoldingPeriod

_MAX_DEDUCTIBLE_LOSS = Decimal("-3000")  # IRS annual cap


def _is_short_term(d) -> bool:
    """Check if a disposal is short-term."""
    hp = d.holding_period
    if isinstance(hp, HoldingPeriod):
        return hp == HoldingPeriod.SHORT_TERM
    return str(hp) in ("short-term", "short", "SHORT_TERM")


def generate_schedule_d_text(
    disposals: list,
    year: int,
    method: str,
    prior_year_carryover: Decimal = Decimal("0"),
) -> str:
    """
    Generate Schedule D summary as a text report.

    Args:
        disposals: list of Disposal dataclass instances
        year: tax year
        method: cost basis method used
        prior_year_carryover: unused loss from prior years (positive number)

    Returns:
        Human-readable Schedule D text
    """
    # Filter to the correct year
    year_disposals = []
    for d in disposals:
        if hasattr(d.date, "year") and d.date.year == year:
            year_disposals.append(d)

    # Compute totals
    st_proceeds = Decimal("0")
    st_basis = Decimal("0")
    st_net = Decimal("0")
    lt_proceeds = Decimal("0")
    lt_basis = Decimal("0")
    lt_net = Decimal("0")

    for d in year_disposals:
        if _is_short_term(d):
            st_proceeds += d.proceeds_usd
            st_basis += d.cost_basis_usd
            st_net += d.gain_loss_usd
        else:
            lt_proceeds += d.proceeds_usd
            lt_basis += d.cost_basis_usd
            lt_net += d.gain_loss_usd

    # Apply prior year carryover
    if prior_year_carryover > Decimal("0"):
        st_net -= prior_year_carryover

    net = st_net + lt_net

    # Deductible loss and carryover
    if net < Decimal("0"):
        deductible = max(net, _MAX_DEDUCTIBLE_LOSS)
        carryover = abs(net - deductible)
    else:
        deductible = Decimal("0")
        carryover = Decimal("0")

    lines = [
        f"Schedule D \u2014 Capital Gains and Losses ({year}, {method})",
        "=" * 60,
        "",
        "PART I \u2014 Short-Term Capital Gains and Losses",
        f"  Proceeds:    ${st_proceeds:>14,.2f}",
        f"  Cost Basis:  ${st_basis:>14,.2f}",
        f"  Net:         ${st_net:>14,.2f}",
        "",
        "PART II \u2014 Long-Term Capital Gains and Losses",
        f"  Proceeds:    ${lt_proceeds:>14,.2f}",
        f"  Cost Basis:  ${lt_basis:>14,.2f}",
        f"  Net:         ${lt_net:>14,.2f}",
        "",
        "PART III \u2014 Summary",
        f"  Net Capital Gain/(Loss):  ${net:>12,.2f}",
    ]
    if net < Decimal("0"):
        lines += [
            f"  Deductible Loss (max $3k): ${deductible:>12,.2f}",
            f"  Carryover to Next Year:   ${carryover:>12,.2f}",
        ]

    return "\n".join(lines)

File: frontend/src/engine/tax/us/schedule-d.ts
/**
 * IRS Schedule D summary generator.
 *
 * Ported from src/tax/us/schedule_d.py.
 *
 * Schedule D aggregates the totals from Form 8949:
 *   - Part I:  Short-term capital gains/losses
 *   - Part II: Long-term capital gains/losses
 *   - Part III: Net capital gain/loss + carryover calculation
 *
 * Capital loss carryover rules:
 *   - Net capital loss limited to $3,000/year deduction against ordinary income
 *   - Excess carries over indefinitely to future years
 *
 * CRITICAL: All monetary values use decimal.js Decimal -- NEVER native number for money.
 */

import Decimal from 'decimal.js';
import type { Disposal, ReportFile } from '../../types';

const MAX_DEDUCTIBLE_LOSS = new Decimal('-3000'); // IRS annual cap

function isShortTerm(d: Disposal): boolean {
  const hp = d.holdingPeriod;
  return ['short-term', 'short', 'SHORT_TERM'].includes(hp);
}

function fmtUsd(d: Decimal): string {
  const num = d.toNumber();
  return num.toLocaleString('en-US', { minimumFractionDigits: 2, maximumFractionDigits: 2 });
}

/**
 * Generate Schedule D summary as a text report.
 *
 * @param disposals          - Array of Disposal objects
 * @param year               - Tax year
 * @param method             - Cost basis method label
 * @param priorYearCarryover - Unused loss from prior years (positive number)
 * @returns                  ReportFile with text content
 */
export function generateScheduleD(
  disposals: Disposal[],
  year: number,
  method: string = 'FIFO',
  priorYearCarryover: Decimal = new Decimal('0'),
): ReportFile {
  // Filter to the correct year
  const yearDisposals = disposals.filter((d) => {
    const dy = parseInt(d.date.slice(0, 4), 10);
    return dy === year;
  });

  let stProceeds = new Decimal('0');
  let stBasis = new Decimal('0');
  let stNet = new Decimal('0');
  let ltProceeds = new Decimal('0');
  let ltBasis = new Decimal('0');
  let ltNet = new Decimal('0');

  for (const d of yearDisposals) {
    if (isShortTerm(d)) {
      stProceeds = stProceeds.plus(d.proceedsUsd);
      stBasis = stBasis.plus(d.costBasisUsd);
      stNet = stNet.plus(d.gainLossUsd);
    } else {
      ltProceeds = ltProceeds.plus(d.proceedsUsd);
      ltBasis = ltBasis.plus(d.costBasisUsd);
      ltNet = ltNet.plus(d.gainLossUsd);
    }
  }

  // Apply prior year carryover
  if (priorYearCarryover.gt('0')) {
    stNet = stNet.minus(priorYearCarryover);
  }

  const net = stNet.plus(ltNet);

  let deductible = new Decimal('0');
  let carryover = new Decimal('0');
  if (net.lt('0')) {
    deductible = Decimal.max(net, MAX_DEDUCTIBLE_LOSS);
    carryover = net.minus(deductible).abs();
  }

  const lines: string[] = [
    `Schedule D \u2014 Capital Gains and Losses (${year}, ${method})`,
    '='.repeat(60),
    '',
    'PART I \u2014 Short-Term Capital Gains and Losses',
    `  Proceeds:    $${fmtUsd(stProceeds).padStart(14)}`,
    `  Cost Basis:  $${fmtUsd(stBasis).padStart(14)}`,
    `  Net:         $${fmtUsd(stNet).padStart(14)}`,
    '',
    'PART II \u2014 Long-Term Capital Gains and Losses',
    `  Proceeds:    $${fmtUsd(ltProceeds).padStart(14)}`,
    `  Cost Basis:  $${fmtUsd(ltBasis).padStart(14)}`,
    `  Net:         $${fmtUsd(ltNet).padStart(14)}`,
    '',
    'PART III \u2014 Summary',
    `  Net Capital Gain/(Loss):  $${fmtUsd(net).padStart(12)}`,
  ];

  if (net.lt('0')) {
    lines.push(`  Deductible Loss (max $3k): $${fmtUsd(deductible).padStart(12)}`);
    lines.push(`  Carryover to Next Year:   $${fmtUsd(carryover).padStart(12)}`);
  }

  return {
    filename: `schedule_d_${year}.txt`,
    content: lines.join('\n'),
    mimeType: 'text/plain',
    reportType: 'schedule_d',
  };
}
